### PR TITLE
Make CLI plot script a proper Python file

### LINKED ISSUE #249: Improve `pineappl plot` Python interaction
Currently, `pineappl plot` relies on Python and Matplotlib. This is actually quite nice, because people using PineAPPL are often familiar with Python, and know how to plot the information, once given.

A current mild pain-point is that there is a lot of string manipulation involved in the call, since a Python script is generated everytime, embedding the data:
https://github.com/NNPDF/pineappl/blob/3476629d40f4b4af164ce067ec2d029781be048c/pineappl_cli/src/plot.rs#L508-L521

However, dumping the script is a design featured, in such a way that people are able to customize it.

I see a few possible ways of improving:
1. split the two `pineappl plot` operations: dump the script and the data separately (in a common data format, like CSV)
2. leverage `pineappl_py` in the script, and directly extract the data from the grid (requires the installation of the PineAPPL python extension)
3. move the plotting feature to `pineappl_py` itself (possibly with a `pineappl-plot` CLI, but it would be installed separately, at least until #176 will be completed)
4. use PyO3 to make the call from the CLI (i.e. from Rust) - in this way people will be able to customize the script and pass it again, but there won't be any need of dumping the data any longer (most likely this is an overkill)

### PR DESCRIPTION
**Pull request by felixhekhorn**
Closes #249 

given the constraints here https://github.com/NNPDF/pineappl/pull/310#issuecomment-2360684605, this is the best we can do - however, I still believe this is a step forward, because at least `plot.py` is now a genuine Python file. It also needs no longer (Rust) formatting, but just a plain find and replace.

@cschwan since you may know the answer straight away (in contrast to me): can you help me with:
https://github.com/NNPDF/pineappl/blob/300f3c65a12c973f9f019c707c9c889dfc606b85/pineappl_cli/src/plot.rs#L549-L550

also note that the plot script itself is badly documented, i.e. there is no documentation, just some matplotlib magic going on, which is hard to understand

I also ran `ruff` on it, which generates more changes

### CODE REVIEW

**Comment by felixhekhorn** [comment]
Actually, after a bit more thinking I believe we can do even more, i.e. almost what I want:
1. I suggest to split all the static stuff, i.e. the plot functions, away from `pineappl_cli/src/plot.py` into a file `pineappl_py/something/plot.py`
2. that new file should be statically included into the `pineappl` Python package (and thus the Python interface will have easy access to the plot functions; we can also provide some additional methods which generate the necessary data from a grid using the Python interface itself)
3. the CLI will copy that file verbatim into the output of `pineappl plot`

File: pineappl_cli/src/plot.py
Comment (felixhekhorn) [reply pineappl_cli/src/plot.py:11]: done in 51a5d536

File: pineappl_cli/src/plot.rs
Comment (felixhekhorn) [inline pineappl_cli/src/plot.rs:567]: I'm almost sure this is not the most efficient way to do things, but it is the only one I managed to make running ...

Comment (alecandido) [reply pineappl_cli/src/plot.rs:569]: Well, it's not completely bad: you're distributing it within the binary.

It's true that in some other respects it may be better to install it somewhere on the file system, and recover it at runtime. But in this way you do not depend at all on the file system and anything else, and the price is increasing the binary size of something that is most likely negligible...
(it will be 10-20KB?)

Comment (felixhekhorn) [reply pineappl_cli/src/plot.rs:569]: nono, I don't mean the idea - it is the best we can do given the constraint (I think) and I will try to implement my idea [from above](https://github.com/NNPDF/pineappl/pull/312#issuecomment-2363034277) to keep the generated file simple and self-contained.

I was more worried of the string manipulation I'm doing here ... I was fighting with `split` and other stuff which did not work

**Comment by cschwan** [comment]
There are a bit too many changes doing different things in here, which make reviewing and fixing the tests difficult. Could you please focus first on one goal, namely making `plot.py` a proper Python file? We can reformat and lint later on when we know it's working. Reformatting will also likely break the NNPDF pheno scripts in the paper repo.

File: pineappl_cli/src/plot.py
Comment (cschwan) [inline pineappl_cli/src/plot.py:24]: Put these booleans together, maybe in a dictionary.

File: pineappl_cli/src/plot.py
Comment (cschwan) [inline pineappl_cli/src/plot.py:89]: All of these `if`s should be put together in a loop.

File: pineappl_cli/src/plot.rs
Comment (cschwan) [inline pineappl_cli/src/plot.rs:127]: This will lead to problems once you use an NNPDF sets; these contain underscores which  will break LaTeX.

**Review by cschwan — CHANGES_REQUESTED** [review]

**Comment by felixhekhorn** [comment]
> There are a bit too many changes doing different things in here, which make reviewing and fixing the tests difficult. Could you please focus first on one goal, namely making `plot.py` a proper Python file? We can reformat and lint later on when we know it's working. Reformatting will also likely break the NNPDF pheno scripts in the paper repo.

mmmmm ... [grummel] ... the commits so far are the minimal changes I would say, i.e. they make `plot.py` a Python file and the only thing on top, which is one of the main motivation of doing this thing, is `ruff` ... I don't know how one would do less

also I want to largely iterate the Python file to make it easier to maintain, to ban all the globals, and to make the plot functions reusable. Actually I already started on the process, but now I don't know whether to commit (and whether my last hour was wasted ...)

File: pineappl_cli/src/plot.py
Comment (felixhekhorn) [reply pineappl_cli/src/plot.py:26]: and kept this! `data` and `metadata` are still at the bottom

File: pineappl_cli/src/plot.py
Comment (felixhekhorn) [reply pineappl_cli/src/plot.py:24]: wait a second: you just told me I should rework so much! and in fact this is a prime example of what I want to rework

File: pineappl_cli/src/plot.py
Comment (felixhekhorn) [reply pineappl_cli/src/plot.py:89]: [see above](https://github.com/NNPDF/pineappl/pull/312#discussion_r1768479509)

File: pineappl_cli/src/plot.rs
Comment (felixhekhorn) [reply pineappl_cli/src/plot.rs:127]: no, I don't think so. 1. this leads to
```
/home/felix/Physik/N3PDF/PineAPPL/pineappl/test-data/plot2.py:525: SyntaxWarning: invalid escape sequence '\_'
  "NNPDF31\_nlo\_as\_0118\_luxqed",
```
and 2. I didn't check the actual plot script yet (since it is too complicated), but this would only be true if the label would be "tex-ified", which I think it does not. I'm testing atm with simple case in the unit test and there the plot (-label) is just fine

**Comment by felixhekhorn** [comment]
So, I committed nevertheless (because I don't want to loose my work).

I would say: if you want to have a smaller PR break this PR at [51a5d53](https://github.com/NNPDF/pineappl/pull/312/commits/51a5d536ae728535f9d2290613f4359b47647e4b). Now that may still be difficult too review, but this is because the previous state was bad - I'm 100% confident that this PR improves the situation. Yes, even that change can and may break the pheno paper plots, but then we can wait until that paper is out to merge - or maybe this PR reaches a state first, where whatever hacks were needed so far there are no longer needed. If you wish we can put the commits after the one mentioned into a separate PR - where I want to change lots of things.

Speaking of questionable content: what exactly is the purpose of `map_format_e_join_repeat_last`? if this is just for getting simpler arguments for matplotlib then I will ban that function and replicate the behaviour on the actual spot (if that is needed - because I believe sometimes it is counteracted again on the Python side)

**Comment by Radonirinaunimi** [comment]
I agree that the previous version of the script was in a lousy state and I think that we should fix this once and for all here. Hence I believe it would be reasonable to expect a few changes. Skimming through I think @felixhekhorn's changes are going in the right direction and it seems that to make the tests pass it will suffice to "copy" the contents of `plot.py` into `tests/plot.rs`.

> may break the pheno paper plots

AFAIA, the pheno paper is not relying on this at all.

PS: I will have to wait for this to resume https://github.com/NNPDF/pineappl/pull/310.

**Comment by alecandido** [comment]
A very stupid and almost irrelevant improvement:
image
```py
ax.set_ylim(0.0)
```
(just to avoid the bottom gap)

**Comment by felixhekhorn** [comment]
> A very stupid and almost irrelevant improvement:

@alecandido Maybe this comment belongs to #309?

**Comment by alecandido** [comment]
> @alecandido Maybe this comment belongs to #309?

Ah, yes, I confused two open tabs, sorry 😅

File: pineappl_cli/src/plot.py
Comment (cschwan) [reply pineappl_cli/src/plot.py:26]: I just tested this and yes, that's indeed the case!

File: pineappl_cli/src/plot.py
Comment (cschwan) [reply pineappl_cli/src/plot.py]: Done in commit 24b97b8cb46fdf5511a94219e7e40e53f74217af.

File: pineappl_cli/src/plot.rs
Comment (cschwan) [reply pineappl_cli/src/plot.rs]: Fixed in commit 68e73907cd8b30148ce3ab00bbee9087b1101971.

File: pineappl_cli/src/plot.py
Comment (felixhekhorn) [reply pineappl_cli/src/plot.py]: Improved in 143b5d53 - `globals()` should be avoided if possible ...

Comment (cschwan) [reply pineappl_cli/src/plot.py]: I believe `globals()` is OK since execute a generated script, but if we want to split parts of it into a library this is better.

**Comment by cschwan** [comment]
Is there anything missing here? If not, I'll fix the tests and merge it into master.

**Comment by Radonirinaunimi** [comment]
> Is there anything missing here? If not, I'll fix the tests and merge it into master.

Not as far as I know. Things LGTM.

**Comment by Radonirinaunimi** [comment]
@felixhekhorn  Can you confirm?

**Comment by felixhekhorn** [comment]
I still want to [copy the file to pineappl_py](https://github.com/NNPDF/pineappl/pull/312#issuecomment-2363034277), but this can also be done later, if you want to merge this quickly (to not block other PRs) (as I'm not sure how quickly I can come back to this).

Then at a later stage, I'd still like to improve the script as a whole and actually now that I'm thinking about it - some part of that will need to take place together with the move, i.e. it is not trivial, so let's postpone ...

**Comment by Radonirinaunimi** [comment]
@cschwan In case you've not worked on this yet I've fixed the tests locally that I could push (if it helps).

**Comment by cschwan** [comment]
Yes, please do!

**Review by Radonirinaunimi — APPROVED** [review]

**Comment by cschwan** [comment]
Are there any immediate concerns that should be addressed? Otherwise I'm going to merge this into `master`.

**Comment by Radonirinaunimi** [comment]
> Are there any immediate concerns that should be addressed? Otherwise I'm going to merge this into `master`.

This actually could be merged.

**Pull request merged**
